tests/test_signals.py

```diff
@@ -21,11 +19,22 @@ def test_churn(self):
         assert np.isclose(churn(self.up, self.up_down), 1)
         assert np.isclose(churn(self.up, self.oscillate), 1)
         assert np.isclose(churn(self.up, self.down), 2)
-        assert np.isnan(churn(self.up, self.constant))
+        self.assertRaisesRegex(
+            AssertionError,
+            "s2 must have non-zero standard deviation",
+            churn,
+            self.up,
+            self.constant,
+        )
 
     def test_churn_tb(self):
-        assert np.isclose(churn(self.up, self.up, top_bottom=2), 0)
-        assert np.isclose(churn(self.up, self.up_down, top_bottom=2), 0.5)
-        assert np.isclose(churn(self.up, self.oscillate, top_bottom=2), 0.5)
-        assert np.isclose(churn(self.up, self.down, top_bottom=2), 1)
-        assert np.isclose(churn(self.up, self.constant, top_bottom=2), 0)
+        tmp = churn(self.up, self.up, top_bottom=2)
+        assert np.isclose(tmp, 0), tmp
+        tmp = churn(self.up, self.up_down, top_bottom=2)
+        assert np.isclose(tmp, 0.5), tmp
```

Comment:
Okay, so I think I've tracked down why this is failing in CI.

First of all I'll note that this doesn't fail locally for me either, *including* when run in docker with the exact same python version and dependency versions - and of course that's under linux. So the only difference I can think exists that makes this *reliably* fail in CI (which it does) while passing locally is the the processor architecture - although actually the error is real and could happen any time locally too.

I notice `numpy.argsort` has gained the `stable=` parameter in version 2.0 - not that you can use that since we claim to support `numpy>=1.26.2`. But that's the hint about what's amis: the sort used by `argsort` is not stable. I'm not sure there's a really reliable way to induce one sort of instability, but changing the sorting algorithm from quicksort to heapsort can give it a nudge to reproduce the issue - here's a patch that will trigger a local failure (for me):

```
Index: numerai_tools/scoring.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/numerai_tools/scoring.py b/numerai_tools/scoring.py
--- a/numerai_tools/scoring.py	(revision 1abdfbec329c41a1765299eef63e32fe60041513)
+++ b/numerai_tools/scoring.py	(date 1726053548703)
@@ -54,7 +54,7 @@
     Returns:
         pd.Series - the filtered and sorted data
     """
-    tb_idx = np.argsort(s)
+    tb_idx = np.argsort(s, kind="heapsort")
     top = s.iloc[tb_idx[:top_bottom]]
     bot = s.iloc[tb_idx[-top_bottom:]]
     if return_concatenated:

```

So CI is right, churn really does sometimes return 0.75 for these inputs (it's all dependent on whether you get element `1` or element `4` from `self.up_down`). I think you have two options:
- We can set some rules about sort order when we look at the tb of the portfolio, and follow those both when ingesting submissions and here. e.g. we can rank with tie-break (which I think we already have an implementatin for)
- Or you can use `np.argsort(..., kind='stable')` which is around for quite old numpy versions and is effectively the same thing as tie-breaking by rank.

A helpful way to think here is that any time you sort without a stable algorithm, you're effectively pre-shuffling the inputs (... except that that shuffle depends on the local environment so won't match between CI and local - or indeed user machines). Yuck.

Reply:
Awesome, thank you for this information. I'll incorporate `kind='stable'` - hopefully this will fix the incongruity with CI.